src/audio_core/hle/wmf_decoder_utils.cpp

```diff
@@ -0,0 +1,366 @@
+    if (FAILED(hr)) {
+        ReportError("Failed to end streaming for MFT", hr);
+    }
+
+    return 0;
+}
+
+int send_sample(IMFTransform* transform, DWORD in_stream_id, IMFSample* in_sample) {
+    HRESULT hr = S_OK;
+
+    if (in_sample) {
+        hr = transform->ProcessInput(in_stream_id, in_sample, 0);
+        if (hr == MF_E_NOTACCEPTING) {
+            return 1; // try again
+        } else if (FAILED(hr)) {
+            ReportError("MFT: Failed to process input", hr);
+            return -1;
+        } // FAILED(hr)
+    } else {
+        hr = transform->ProcessMessage(MFT_MESSAGE_COMMAND_DRAIN, 0);
+        // ffmpeg: Some MFTs (AC3) will send a frame after each drain command (???), so
+        // ffmpeg: this is required to make draining actually terminate.
+        if (FAILED(hr)) {
+            ReportError("MFT: Failed to drain when processing input", hr);
+        }
+    }
+
+    return 0;
+}
+
+// return: 0: okay; 1: needs more sample; 2: needs reconfiguring; 3: more data available
+int receive_sample(IMFTransform* transform, DWORD out_stream_id, IMFSample** out_sample) {
+    HRESULT hr;
+    MFT_OUTPUT_DATA_BUFFER out_buffers;
+    IMFSample* sample = NULL;
+    MFT_OUTPUT_STREAM_INFO out_info;
+    DWORD status = 0;
+    bool mft_create_sample = false;
+
+    if (!out_sample) {
+        ReportError("NULL pointer passed to receive_sample()", MF_E_SAMPLE_NOT_WRITABLE);
+        return -1;
+    }
+
+    hr = transform->GetOutputStreamInfo(out_stream_id, &out_info);
+
+    if (FAILED(hr)) {
+        ReportError("MFT: Failed to get stream info", hr);
+        return -1;
+    }
+    mft_create_sample = (out_info.dwFlags & MFT_OUTPUT_STREAM_PROVIDES_SAMPLES) ||
+                        (out_info.dwFlags & MFT_OUTPUT_STREAM_CAN_PROVIDE_SAMPLES);
+
+    while (true) {
+        sample = NULL;
+        *out_sample = NULL;
+        status = 0;
+
+        if (!mft_create_sample) {
+            sample = create_sample(NULL, out_info.cbSize, out_info.cbAlignment);
+            if (!sample) {
+                ReportError("MFT: Unable to allocate memory for samples", hr);
+                return -1;
+            }
+        }
+
+        out_buffers.dwStreamID = out_stream_id;
+        out_buffers.pSample = sample;
+
+        hr = transform->ProcessOutput(0, 1, &out_buffers, &status);
+
+        if (!FAILED(hr)) {
+            *out_sample = out_buffers.pSample;
+            break;
+        }
+
+        if (hr == MF_E_TRANSFORM_NEED_MORE_INPUT) {
+            // TODO: better handling try again and EOF cases using drain value
+            return 1;
+        }
+
+        if (hr == MF_E_TRANSFORM_STREAM_CHANGE) {
+            ReportError("MFT: stream format changed, re-configuration required", hr);
+            return 2;
+        }
+
+        break;
+    }
+
+    if (out_buffers.dwStatus & MFT_OUTPUT_DATA_BUFFER_INCOMPLETE) {
+        return 3;
+    }
+
+    // TODO: better handling try again and EOF cases using drain value
```

Comment:
TODO(liushuyu):

Reply:
I just removed the comment since AAC decoder is a `SYNCMFT` it does not need to keep track of drain status if you are doing the drain "correctly"